### PR TITLE
Port applicable findings from LLM bug audit to upstream repo (plus other tweaks)

### PR DESCRIPTION
**Pull request by t0mnt**
Ran an LLM audit of my repo before merging a big branch and some stuff was more applicable here than there. Feel free to keep only select parts or reject it all but there's plenty in here that seemed useful to me 

LLM Description:

Physics / features:
- embedding.py: compute tagging features BEFORE the jet-rest-frame boost. The old boost-then-features order degenerated 4/7 features (rest-frame jet pt~0 -> clamped pt_jet, atan2/eta of a ~zero vector) and measurably broke even SO(2) invariance of every learned-frames row (xyrotation max MSE O(1e2..1e4) -> O(1e-8); the newly added assertion in test_tag_invariance catches it). mass_reg no longer converts the spacelike beam spurion into a lightlike one.
- wrappers.py: ParticleNetWrapper (phi,eta) kNN index offset by extra_scalars (the hardcoded [4,5] was correct only for top-tagging; on JetClass it clustered layer-0 kNN by PID one-hots).

Finetune path:
- experiment.py: _init_optimizer respects caller-supplied param_groups (the finetune backbone/head lrs were being clobbered by the ParT name match).
- finetuneexperiment.py: read cfg.ema_decay (top level), not the nonexistent cfg.training.ema_decay (struct-mode crash for ema=true backbones).

Safety nets:
- test_tag_invariance.py: assert (max MSE < 1e-5) instead of print-and-return, and transform each jet independently (catches cross-jet leakage).
- test_tag_flops.py: pytest.skip env-dependent init failures instead of silent PASS.

Robustness:
- base_experiment.py: scaler.update() before the max_grad_norm skip-return, so an AMP overflow can shrink the loss scale instead of freezing it (which silently skipped every later iteration).

### CODE REVIEW

**Comment by t0mnt** [comment]
Does trimmer run after prepare frames for the lloca library ParT that this repo calls? that too looks like a bug
not sure where to comment but this seems as good a place as any

**Comment by t0mnt** [comment]
I also have a changed `get_datasets.py` that fetches `TopTaggingXL` and/or `JetClass`
Let me know if you feel that's meaningful enough to be added here, in a separate PR, or not at all

**Comment by t0mnt** [comment]
why particlenet `fusion` false? and speaking of weaver defaults, could have `deta = (eta_4 - eta_jet) * torch.sign(eta_jet)`

**Comment by t0mnt** [comment]
maybe this stuff will be valuable for the new repo? I saw it mentioned above in some of your presentations and the like but if I new it was coming I'd have tried to sink less cost into this repo :sob:

ps: `test_tag_invariance.py` deleted without explanation (from the `.yaml` and now explained, EDIT)

**Comment by t0mnt** [comment]
> I also have a changed `get_datasets.py` that fetches `TopTaggingXL` and/or `JetClass` Let me know if you feel that's meaningful enough to be added here, in a separate PR, or not at all

It also replaces wget with requests (supports retrys, is what the official zenodo download tool is built off of, and yknow isnt abandonware)

plus timestamping downloads but that's minor

Anyways after I update my repo to support L-GATr 2.0, and `torch.compile()` support for CGENN & maybe some others it's probably time to defork (credits will still be clear in license, readme, git history, etc)
so I'll do a last ping soon if anything here is worth keeping before `lloca-experiments` is officially frozen/archived/whatever

at the very least the change in `requirements.txt` of `lgatr[xformers-attention]>=1.4.2, <2.0.0` is pretty required

**Comment by t0mnt** [comment]
@spinjo see above, many changes here would be very helpful as a final fix before frozen or if definitively frozen, applicable to the new tagging environment (including stuff not in the PR for the latter)

just give it a glance :eyes:

**Comment by t0mnt** [comment]
Don't want to re-ping but I'm finished and would appreciate a look (will defork end of week)






EDIT:
All relevant changes have been, everything else I would make personal preference that you won't go for (like training with the ParticleNet in /baselines and mortifying it instead of calling the lloca library for a backbone)
Let me know if any changes might be merged, if nothing (not even `lgatr <2.0`) I can at least close the PR

**Comment by spinjo** [comment]
Sorry, I've been busy finishing the https://github.com/heidelberg-hepml/tagging-guide repo and then on vacation. Will take care of this in the next days. I'm definitely planning to merge it, just want to take the time to go through it carefully.

File: config_quick/jctagging.yaml [hunk 1]
```diff
@@ -2,12 +2,16 @@ exp_type: jctagging
 exp_name: jc_debug
 
 data:
- data_dir: data/jetclass/
+ data_dir: data/JetClass/Pythia
  features: default
  data_config: null
- train_files_range: [122,123]
```
Comment (spinjo) [inline config_quick/jctagging.yaml:8]: The `[122,123]` etc trick was what I used on my local device, in the end it doesn't matter which files are used for testing because they are all from the same simulation. The `tagging-guide` repo has a cleaner solution. So I would prefer to just stick to the old solution, and remove the comment. Same thing for the `toptagxl` file.

File: experiments/baselines/cgenn/mvlayernorm.py [hunk 2]
```diff
@@ -12,9 +12,14 @@ def __init__(self, algebra, channels):
         super().__init__()
         self.algebra = algebra
         self.channels = channels
-        self.a = nn.Parameter(torch.ones(1, channels))
+        # 1-d (channels,) rather than the official (1, channels): a norm GAIN must fall
```
Comment (spinjo) [inline experiments/baselines/cgenn/mvlayernorm.py]: Agreed on the fix, but please remove the comment

File: experiments/tagging/miniweaver/dataset.py [hunk 3]
```diff
@@ -247,7 +247,11 @@ def restart(self):
             self.load_filelist_and_ranges = [
                 (
                     self.filelist[i : i + self._fetch_step],
-                    [self.load_range] * self._fetch_step,
```
Comment (spinjo) [inline experiments/tagging/miniweaver/dataset.py:250]: I wouldn't change this because we didn't require it for our paper with the datasets that we had back then. The tagging-guide repo has a solution for this.

File: experiments/tagging/miniweaver/dataset.py [hunk 4]
```diff
@@ -351,11 +355,11 @@ def _try_get_next(self, init=False):
         end_of_list = self.ipos >= len(self.load_filelist_and_ranges)
         if end_of_list:
             if init:
-                raise RuntimeError(
```
Comment (spinjo) [inline experiments/tagging/miniweaver/dataset.py:354]: This is just logging so I would undo the change

File: experiments/tagging/miniweaver/fileio.py [hunk 5]
```diff
@@ -133,6 +133,15 @@ def _read_files(
             elif ext == ".parquet":
                 a = _read_parquet(filepath, branches, load_range=load_range)
         except Exception as e:
+            # weaver proper logs the error + traceback here; a bare `a = None` swallowed
```
Comment (spinjo) [inline experiments/tagging/miniweaver/fileio.py]: This is just logging so I would undo the change

File: experiments/tagging/miniweaver/loader.py
Comment (spinjo) [inline experiments/tagging/miniweaver/loader.py:1]: this is just logging so I would undo the change

File: experiments/tagging/embedding.py [hunk 6]
```diff
@@ -102,22 +102,23 @@ def embed_tagging_data(fourmomenta, scalars, ptr, cfg_data):
...
 
-    if cfg_data.boost_jet:
-        # boost to the jet rest frame to avoid large boosts
-        jet = scatter(
-            fourmomenta[~is_spurion], batch[~is_spurion], dim=0, reduce="sum"
-        ).index_select(0, batch)
-        jet_boost = restframe_boost(jet)
-        fourmomenta = torch.einsum("ijk,ik->ij", jet_boost, fourmomenta)
-
+    # tagging features are computed BEFORE any jet-rest-frame boost, from the original
```
Comment (spinjo) [inline experiments/tagging/embedding.py]: Agreed on the change, but please remove the comment. This is also fixed in the tagging-guide repo.

File: experiments/tagging/embedding.py [hunk 7]
```diff
@@ -102,22 +102,23 @@ def embed_tagging_data(fourmomenta, scalars, ptr, cfg_data):
         scalars[~is_spurion] = scalars_buffer
         ptr[1:] = ptr[1:] + (arange + 1) * n_spurions
 
-    # add mass regulator
+    # add mass regulator -- to the REAL particles only: a spacelike beam spurion
```
Comment (spinjo) [inline experiments/tagging/embedding.py]: Agreed on the change, but please restore the comment to the original (short) form

File: experiments/tagging/experiment.py [hunk 8]
```diff
@@ -78,6 +78,28 @@ def init_physics(self):
                 )
                 self.cfg.model.framesnet.equivectors.num_scalars = self.extra_scalars
                 self.cfg.model.framesnet.equivectors.num_scalars += num_tagging_features
+                # boost_jet lands each jet at rest, and PURE-ROTATION frame families cannot
```
Comment (spinjo) [inline experiments/tagging/experiment.py]: This is a fair point, although its a bit of a exotic combination. I'm happy to keep it, but please remove the comment.

File: experiments/tagging/experiment.py [hunk 9]
```diff
@@ -160,7 +182,9 @@ def init_standardization(self):
             self.model.init_standardization(embedding["fourmomenta"], embedding["ptr"])
 
     def _init_optimizer(self, param_groups=None):
-        if self.cfg.model.net._target_.rsplit(".", 1)[-1] in [
```
Comment (spinjo) [inline experiments/tagging/experiment.py:163]: Fair change, happy to keep it, but please remove the comment

File: experiments/tagging/finetuneexperiment.py [hunk 10]
```diff
@@ -121,8 +121,11 @@ def init_model(self):
 
         if self.cfg.ema:
             LOGGER.info("Re-initializing EMA")
+            # NB: ema_decay lives at the config TOP level (config/default.yaml), not under
```
Comment (spinjo) [inline experiments/tagging/finetuneexperiment.py]: Good point, but please remove the comment

File: experiments/tagging/jetclassexperiment.py [hunk 11]
```diff
@@ -80,9 +80,12 @@ def init_data(self):
                 for classname in self.class_names
                 for i in range(*files_range[label])
             ]
-            file_dict, _ = to_filelist(flist)
```
Comment (spinjo) [inline experiments/tagging/jetclassexperiment.py:83]: This is logging so I would remove it.

File: experiments/tagging/jetclassexperiment.py [hunk 12]
```diff
@@ -97,7 +100,10 @@ def init_data(self):
                 file_fraction=1,
                 fetch_by_files=self.cfg.jc_params.fetch_by_files,
                 fetch_step=self.cfg.jc_params.fetch_step,
-                infinity_mode=self.cfg.jc_params.steps_per_epoch is not None,
```
Comment (spinjo) [inline experiments/tagging/jetclassexperiment.py:100]: My understanding is that this catches a special edge case where people set `steps_per_epoch` (we don't do that) so I would propose to undo your change.

File: experiments/tagging/toptagxlexperiment.py
Comment (spinjo) [inline experiments/tagging/toptagxlexperiment.py:1]: see `jetclassexperiment.py` comments

File: experiments/base_experiment.py [hunk 13]
```diff
@@ -331,27 +331,35 @@ def _init_backend(self):
     def _init_optimizer(self, param_groups=None):
         if param_groups is None:
 
-            def is_bias(param):
```
Comment (spinjo) [inline experiments/base_experiment.py:334]: Happy to keep this, but please remove the comment

File: experiments/base_experiment.py [hunk 14]
```diff
@@ -527,6 +535,7 @@ def train(self):
 
         # early stopping
         smallest_val_loss, smallest_val_loss_step = 1e10, 0
+        self._best_state = None  # in-RAM best-val checkpoint, used only when save=False
```
Comment (spinjo) [inline experiments/base_experiment.py]: I would undo this because `save=False` is only meant to be used for debugging

File: .github/workflows/experiments_amplitudes.yaml
Comment (spinjo) [inline .github/workflows/experiments_amplitudes.yaml:1]: Please undo these changes - the github workflows were just for our debugging and we didn't maintain them properly during the project (we even stopped using them halfway because we ran out of credits or so). We could also just delete them. Also holds for the changes in the other 3 files.

File: experiments/baselines/mipart.py [hunk 15]
```diff
@@ -1204,7 +1204,11 @@ def forward(self, x, v=None, mask=None, uu=None, uu_idx=None):
             # print(x_cls,x_cls.shape)
             output = self.fc(x_cls)
             if self.for_inference:
-                output = torch.softmax(output, dim=1)
```
Comment (spinjo) [inline experiments/baselines/mipart.py:1207]: Fair point, but please remove the comments. Same for ParticleNet and ParT below.

Side note: The weaver env that mipart, part and particlenet were developed for predicts (batchsize,2) labels instead of the (batchsize,1) labels that we have.

File: tests/experiments/test_tag_flops.py [hunk 16]
```diff
@@ -52,8 +52,10 @@ def test_tagging(framesnet, model_list, equivectors, jet_size=50):
     exp.init_physics()
     try:
         exp.init_model()
-    except Exception:
-        return
+    except Exception as e:
```
Comment (spinjo) [inline tests/experiments/test_tag_flops.py:55]: This is fine, but please remove the comment

File: experiments/tagging/wrappers.py [hunk 17]
```diff
@@ -339,8 +339,16 @@ def forward(self, embedding):
             batch,
             tracker,
         ) = super().forward(embedding)
-        # ParticleNet uses L2 norm in (phi, eta) for kNN
```
Comment (spinjo) [inline experiments/tagging/wrappers.py:342]: This is fine, but please remove the comment

File: tests/experiments/test_tag_invariance.py
Comment (spinjo) [inline tests/experiments/test_tag_invariance.py:1]: This is fine, but please remove the comments

**Review by spinjo — CHANGES_REQUESTED** [review]
Thanks for including the issues that you found during your work, thats great practice!

I went through all the changes and left comments where things should be changed before merging. The thoughts are:
- There were some real bugs that you found - thanks! But as far as I can tell, they are also fixed in https://github.com/heidelberg-hepml/tagging-guide.
- The previous code tried to be minimal wrt comments. You added comments for all things you fixed. This is great for me reviewing it, but once its part of the default code its annoying because the bug is already fixed - so its not so relevant to explain the fix. Or otherwise the code eventually contains an insane amount of comments. Thats why I often asked to remove them.
- You often extended logging. This is fair, but for this purpose here I'd prefer the changes to be minimal, so I would remove them.
- Your fixes sometimes concern special cases that were not intended to be triggered with the code base. So I thought about it case by case.

If you want this to be merged soon, you can just go ahead and implement the changes yourself.

Thanks!

**Comment by t0mnt** [comment]
Making the changes, skim through the above comments if you'd like, and I'd appreciate a list of `tagging-guide` improvements that I can port to my codebase if there's more beyond the 3 you said, which there likely is (though this probably isn't exactly trivial)

File: experiments/tagging/embedding.py (hunk 6 above)
Comment (t0mnt) [reply experiments/tagging/embedding.py]: Is it?

**Review by spinjo — APPROVED** [review]
Thanks for the edits, ready to merge from my side. Give a thumbs up if you agree and I'll merge!

**Comment by t0mnt** [comment]
> I also have a changed `get_datasets.py` that fetches `TopTaggingXL` and/or `JetClass` Let me know if you feel that's meaningful enough to be added here, in a separate PR, or not at all

Glance at some of the comments I left and then sure, I'm ready to merge after this next commit 😃

**Comment by t0mnt** [comment]
> I also have a changed `get_datasets.py` that fetches `TopTaggingXL` and/or `JetClass` Let me know if you feel that's meaningful enough to be added here, in a separate PR, or not at all

?
also for `embedding.py` the ptr gets mutated

**Comment by t0mnt** [comment]
@spinjo hey did you reproduce LorentzNet on regular top landscape? Because I just did for 3 trials and got 93.93%±0.0002 which is like 14σ off. In the config the scheduler differs but I can't imagine it's that extreme, and `Reproduce.md` only finds LorentzNet on XL and JetClass.

**Comment by t0mnt** [comment]
If I had to guess after looking at the epochs it's probably the scheduler

**Comment by t0mnt** [comment]
and is there a jc_slim config? I trained on my default gt config and got 86.54%, about the same as the published 86.6% (maybe equal if you rounded). Is it fair to use that number and/or is a config missing (maybe jc_transformer used)?

**Comment by spinjo** [comment]
I think its fine to keep `get_datasets.py` as it is. 

Concerning LorentzNet, I agree that its probably the scheduler. We didnt validate on toptagging because of the ugly scheduler, we only did JetClass there.

Concerning LGATrSlim on JetClass, we used jc_ParT for all models except the original LGATr in our publications.

**Comment by t0mnt** [comment]
I got CGENN up to 94.25% accuracy by running my basic graph transformer config (just AdamW on ParticleNet/ParT's epoch schedule with lr 1e-3 and weight decay 0.01) by accident, training in "only" 10 hours on Top Landscape. This would be a hefty port here or `tagging-guide` since I give it compile support, sparse GP, and even triton kernels using Kingdon. But since it basically already was on this repo (previous fork) it wouldn't be too crazy to add here or elsewhere. Let me know if there's interest in that work.
